sound/soc/sof/ipc4-pcm.c

```diff
@@ -32,65 +60,184 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+	pipeline_widget = swidget->pipe_widget;
+	pipeline = pipeline_widget->private;
+
+	/* do not add backend pipelines */
+	if (pipeline->is_backend)
+		goto src_add;
+
+	/* check if pipeline has already been added to the list */
+	for (i = 0 ; i < pipeline_info->count; i++)
+		if (pipeline_info->pipelines[i] == pipeline_widget)
+			goto src_add;
+
+	pipeline_count = pipeline_info->count++;
+	if (pipeline_info->count > SOF_IPC4_TRIGGER_MAX_PIPELINE_COUNT) {
+		dev_err(sdev->dev, "Cannot trigger %d pipelines at once\n", pipeline_info->count);
+		return -EINVAL;
+	}
+
+	pipeline_info->pipelines[pipeline_count] = pipeline_widget;
+
+src_add:
+	/* add pipelines in the source paths */
+	snd_soc_dapm_widget_for_each_source_path(widget, p) {
+		if (!widget_in_list(list, p->source))
+			continue;
+
+		if (!p->walking && p->source->dobj.private) {
+			p->walking = true;
+			ret = sof_ipc4_get_pipelines_in_order(sdev, p->source, list, pipeline_info);
+			p->walking = false;
+			if (ret < 0)
+				return ret;
+		}
+	}
+
+	return 0;
+}
+
 static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 				      struct snd_pcm_substream *substream, int state)
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
 	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct sof_ipc4_multi_pipeline_info pipeline_info;
 	struct snd_sof_widget *pipeline_widget;
 	struct snd_soc_dapm_widget_list *list;
 	struct snd_soc_dapm_widget *widget;
 	struct sof_ipc4_pipeline *pipeline;
 	struct snd_sof_widget *swidget;
 	struct snd_sof_pcm *spcm;
 	int ret = 0;
-	int num_widgets;
+	int num_widgets, i;
 
 	spcm = snd_sof_find_spcm_dai(component, rtd);
 	if (!spcm)
 		return -EINVAL;
 
 	list = spcm->stream[substream->stream].list;
 
+	pipeline_info.count = 0;
+
+	/* get the list of pipeline widgets that need to be triggered */
 	for_each_dapm_widgets(list, num_widgets, widget) {
 		swidget = widget->dobj.private;
 
 		if (!swidget)
 			continue;
 
-		/* find pipeline widget for the pipeline that this widget belongs to */
-		pipeline_widget = swidget->pipe_widget;
-		pipeline = pipeline_widget->private;
+		/*
+		 * IPC4 requires pipelines to be triggered in order starting at the sink and
+		 * walking all the way to the source. So the starting widget for playback is of
+		 * DAI type and for capture, it is of AIF type.
+		 */
+		if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK && !WIDGET_IS_DAI(widget->id))
+			continue;
 
-		/* skip triggering backend pipelines */
-		if (pipeline->is_backend || pipeline->state == state)
+		if (substream->stream == SNDRV_PCM_STREAM_CAPTURE && !WIDGET_IS_AIF(widget->id))
 			continue;
 
+		ret = sof_ipc4_get_pipelines_in_order(sdev, widget, list, &pipeline_info);
+		if (ret < 0)
+			return ret;
+	}
+
+	if (!pipeline_info.count)
+		return ret;
+
+	/* trigger a single pipeline if count is 1 */
+	if (pipeline_info.count == 1) {
```

Comment:
and that single pipeline is the FE pipeline?

Reply:
yes